### ISSUE TITLE
Scatterplot with color for categorical value outputs attribute error about legend

### ISSUE THREAD

**Issue by albert239825** [scverse/scanpy#3102]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of scanpy.
- [X] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

When I try to plot a scatter plot color coded by categorical data, it still output the image, but the legend does not include colors. Additionally, if i use continuous data as the key for plotting, the code executes as expected. Thanks!

![image](https://github.com/scverse/scanpy/assets/43973217/8c1b0a03-3c0b-4452-b759-9cf588b45c53)

### Minimal code sample

```python
sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts', color='reference', show=True)
```

### Error output

```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[37], line 1
----> 1 sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts', color='reference', show=True)

File /project/hipaa_ycheng11lab/atlas/CAMR2024/py311env/lib/python3.11/site-packages/legacy_api_wrap/__init__.py:80, in legacy_api.<locals>.wrapper.<locals>.fn_compatible(*args_all, **kw)
     77 @wraps(fn)
     78 def fn_compatible(*args_all: P.args, **kw: P.kwargs) -> R:
     79     if len(args_all) <= n_positional:
---> 80         return fn(*args_all, **kw)
     82     args_pos: P.args
     83     args_pos, args_rest = args_all[:n_positional], args_all[n_positional:]

File /project/hipaa_ycheng11lab/atlas/CAMR2024/py311env/lib/python3.11/site-packages/scanpy/plotting/_anndata.py:166, in scatter(adata, x, y, color, use_raw, layers, sort_order, alpha, basis, groups, components, projection, legend_loc, legend_fontsize, legend_fontweight, legend_fontoutline, color_map, palette, frameon, right_margin, left_margin, size, marker, title, show, save, ax)
    160     raise ValueError("Either provide a `basis` or `x` and `y`.")
    161 if (
    162     (x in adata.obs.keys() or x in var_index)
    163     and (y in adata.obs.keys() or y in var_index)
    164     and (color is None or color in adata.obs.keys() or color in var_index)
    165 ):
--> 166     return _scatter_obs(**args)
    167 if (
    168     (x in adata.var.keys() or x in adata.obs.index)
    169     and (y in adata.var.keys() or y in adata.obs.index)
    170     and (color is None or color in adata.var.keys() or color in adata.obs.index)
    171 ):
    172     adata_T = adata.T

File /project/hipaa_ycheng11lab/atlas/CAMR2024/py311env/lib/python3.11/site-packages/scanpy/plotting/_anndata.py:521, in _scatter_obs(adata, x, y, color, use_raw, layers, sort_order, alpha, basis, groups, components, projection, legend_loc, legend_fontsize, legend_fontweight, legend_fontoutline, color_map, palette, frameon, right_margin, left_margin, size, marker, title, show, save, ax)
    517         legend = axs[ikey].legend(
    518             frameon=False, loc=legend_loc, fontsize=legend_fontsize
    519         )
    520     if legend is not None:
--> 521         for handle in legend.legendHandles:
    522             handle.set_sizes([300.0])
    524 # draw a frame around the scatter

AttributeError: 'Legend' object has no attribute 'legendHandles'
```

### Versions

<details>

```
-----
anndata     0.10.7
scanpy      1.10.1
-----
PIL                 10.3.0
cffi                1.16.0
cycler              0.12.1
cython_runtime      NA
dateutil            2.9.0.post0
defusedxml          0.7.1
dill                0.3.8
h5py                3.11.0
joblib              1.4.2
kiwisolver          1.4.5
legacy_api_wrap     NA
llvmlite            0.42.0
matplotlib          3.9.0
mpl_toolkits        NA
natsort             8.4.0
numba               0.59.1
numexpr             2.10.0
numpy               1.26.4
packaging           24.0
pandas              2.2.2
psutil              5.9.8
pyparsing           3.1.2
pytz                2024.1
scipy               1.13.1
session_info        1.0.0
six                 1.16.0
sklearn             1.5.0
threadpoolctl       3.5.0
torch               2.3.1+cu121
torchgen            NA
tqdm                4.66.4
typing_extensions   NA
yaml                6.0.1
-----
Python 3.11.7 (main, Jan 17 2024, 16:00:28) [GCC 8.5.0]
Linux-3.10.0-1160.114.2.el7.x86_64-x86_64-with-glibc2.17
-----
Session information updated at 2024-06-06 11:18
```

</details>

**Comment by flying-sheep** [comment]
So there is both a plot and error output? That would explain why not all is shown …

**Comment by albert239825** [comment]
@flying-sheep That is correct. There is both an error and the output of the figure without colors in the legend

**Comment by albert239825** [comment]
I forgot to mention, additionally the same code executes with no problem with older versions of the packages. I've tested it on anndata = 0.8.0 and scanpy 1.9.3 and it executes fine.

**Comment by bio-la** [comment]
Hi @flying-sheep! is there a fix for this issue? (besides pinning scanpy version) 
as i understand it, matplotlib went back to legend_handles ([API](https://matplotlib.org/stable/api/prev_api_changes/api_changes_3.9.0.html#:~:text=Legend.legendHandles%20was%20undocumented%20and%20has%20been%20renamed%20to%20legend_handles.)) and now the legacy wrapper is breaking.

**Comment by flying-sheep** [comment]
The fix was done in #2999, but we didn’t release 1.10.2 yet. Does installing scanpy from git work for you?

**Comment by albert239825** [comment]
@flying-sheep Yes! That fixed it. Thanks!
